python/prairielearn.py

```diff
@@ -52,6 +54,74 @@ class QuestionData(TypedDict):
 class ElementTestData(QuestionData):
     test_type: Literal['correct', 'incorrect', 'invalid']
 
+class DisplayType(Enum):
+    INLINE = "inline"
+    BLOCK = "block"
+
+def grade_question_parameterized(data: QuestionData,
+                                 question_name: str,
+                                 grade_function: Callable[[Any], Tuple[Union[bool, float], Optional[str]]],
+                                 weight: int = 1) -> None:
+    '''
+    Grade question question_name. grade_function should take in a single parameter
+    (which will be the submitted answer) and return a 2-tuple:
+        - The first element of the 2-tuple should either be:
+            - a boolean indicating whether the question should be marked correct
+            - a partial score between 0 and 1, inclusive
+        - The second element of the 2-tuple should either be:
+            - a string containing feedback
+            - None, if there is no feedback (usually this should only occur if the answer is correct)
+    '''
+
+    # Create the data dictionary at first
+    data['partial_scores'][question_name] = {
+        'score': 0.0,
+        'weight': weight
+    }
+
+    try:
+        submitted_answer = data['submitted_answers'][question_name]
+    except KeyError:
+        # Catch error if no answer submitted
+        data["format_errors"][question_name] = 'No answer was submitted'
+        return
+
+    try:
+        result, feedback_content = grade_function(submitted_answer)
+
+        if isinstance(result, bool):
+            partial_score = 1.0 if result else 0.0
+        elif isinstance(result, (float, int)):
+            assert 0.0 <= result <= 1.0
+            partial_score = result
+        else:
+            assert_never(result)
+
+    except ValueError as err:
+        data["format_errors"][question_name] = str(err)
+        return
+
+
+    data['partial_scores'][question_name]['score'] = partial_score
+
+    if feedback_content:
+        data['partial_scores'][question_name]['feedback'] = feedback_content
+
+def determine_score_params(score: Optional[float]) -> Tuple[str, float]:
```

Comment:
Same question for this function: if it's only used in `pl-big-o-input`, let's not add it here.